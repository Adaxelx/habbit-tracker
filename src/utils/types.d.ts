export type GridProps = {
  events: Event[];
  month: number;
  year: number;
  moveDate: Function;
};

export interface Event {
  _id: string;
  daysOfWeek: number[];
  title: string;
  timeStart?: string;
  timeEnd?: string;
  dateStart: string;
  dateEnd: string;
  label?: string | Label;
  description?: string;
  userId: string;
  checked?: boolean;
}

export interface EventLoop extends Event {
  numericStart: number;
  numericEnd: number;
}

export type Label = {
  _id: string;
  title: string;
  color: string;
};

export type CalendarTile = {
  day: number;
  events: EventLoop[];
  id: number;
};

export enum SIDES {
  LEFT = -1,
  RIGHT = 1,
}
