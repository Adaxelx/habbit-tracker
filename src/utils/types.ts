export type GridProps = {
  events: Event[];
  month: number;
  year: number;
  moveDate: Function;
  handleDayChange: Function;
};

export interface Event {
  _id: string;
  daysOfWeek: number[];
  title: string;
  timeStart?: string;
  timeEnd?: string;
  dateStart: string;
  dateEnd: string;
  label?: string | Label;
  description?: string;
  userId: string;
  checked?: [
    {
      day: number;
      month: number;
      year: number;
      _id: string;
    },
  ];
}

export interface EventLoop extends Event {
  numericStart: number;
  numericEnd: number;
}

export type Label = {
  _id: string;
  title: string;
  color: string;
  userId: string;
};

export type CalendarTile = {
  day: number;
  events: EventLoop[];
  id: number;
};

export enum SIDES {
  LEFT = -1,
  RIGHT = 1,
}

export type Color = { color?: string; id?: string };

export type NavProps = {
  header: string;
  moveDate: Function;
};

export interface DayCardProps {
  day: string;
  token: string | undefined;
}
