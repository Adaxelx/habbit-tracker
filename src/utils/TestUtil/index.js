export { default } from './TestUtil.ts';
