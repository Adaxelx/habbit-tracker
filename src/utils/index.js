export { default as TestUtil } from './TestUtil';
export { default as RenderWithRouter } from './RenderWithRouter';
export * from './validators';
export * from './types.d';
