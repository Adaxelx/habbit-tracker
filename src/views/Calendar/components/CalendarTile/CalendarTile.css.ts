import styled from 'styled-components';

export const StyledTile = styled.div`
  width: ${({ theme }) => `${theme.gridTile.size}px`};
  height: ${({ theme }) => `${theme.gridTile.size}px`};

  position: relative;
  margin-top: 20%;
  overflow: hidden;
`;

type CirlcleProps = {
  size: number;
  alone: boolean;
  position: number;
  color: string;
};

export const StyledCircle = styled.div<CirlcleProps>`
  position: absolute;
  border-radius: 50%;
  width: ${({ size }) => `${size}px`};
  height: ${({ size }) => `${size}px`};
  background-color: ${({ color }) => color};
  top: ${({ position }) => `${position}px`};
  left: ${({ position }) => `${position}px`};
  ${({ alone }) => alone && 'transform: translate(-50%,-50%);top: 50%;left:50%'}
`;

export const StyledDay = styled.p`
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  z-index: ${({ theme }) => theme.zIndex.medium};
`;
