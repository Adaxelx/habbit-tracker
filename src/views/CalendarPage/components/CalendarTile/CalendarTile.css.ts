import styled from 'styled-components';

export const StyledTile = styled.div`
  width: ${({ theme }) => `${theme.gridTile.size}px`};
  height: ${({ theme }) => `${theme.gridTile.size}px`};
  position: relative;
  overflow: hidden;
  background-color: ${({ theme }) => theme.colors.common.white};
  @media (min-width: ${({ theme }) => theme.media.tablet.s}) {
    width: ${({ theme }) => `${theme.gridTile.size * 2}px`};
    height: ${({ theme }) => `${theme.gridTile.size * 2}px`};
  }
  &:hover {
    cursor: pointer;
  }
`;

type CirlcleProps = {
  size: number;
  alone: boolean;
  position: number;
  color: string | undefined;
};

export const StyledCircle = styled.div<CirlcleProps>`
  position: absolute;
  border-radius: 50%;
  width: ${({ size, alone }) => (alone ? `${size / 1.3}px` : `${size}px`)};
  height: ${({ size, alone }) => (alone ? `${size / 1.3}px` : `${size}px`)};
  background-color: ${({ color }) => color};
  top: ${({ position }) => `${position}px`};
  left: ${({ position }) => `${position}px`};
  ${({ alone }) => alone && 'transform: translate(-50%,-50%);top: 50%;left:50%'}
`;

export const StyledDay = styled.p`
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  z-index: ${({ theme }) => theme.zIndex.low};
`;
