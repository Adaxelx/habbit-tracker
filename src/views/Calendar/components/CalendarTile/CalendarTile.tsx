import React from 'react';
import { StyledCircle, StyledTile } from './CalendarTile.css';

const generateCircles = (colors: string[]) => {
  const size = (50 * 1.3) / colors.length;
  return colors.map((color) => <StyledCircle key={color} color={color} size={size} />);
};

type TileProps = {
  colors: string[];
};

const CalendarTile = ({ colors }: TileProps) => <StyledTile>{generateCircles(colors)}</StyledTile>;

export default CalendarTile;
