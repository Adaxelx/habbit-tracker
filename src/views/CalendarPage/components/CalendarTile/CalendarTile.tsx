import React, { MouseEventHandler } from 'react';
import { useThemeContext } from 'context';
import { Color } from 'utils';
import { useWindowSize } from 'hooks';
import { StyledCircle, StyledTile, StyledDay } from './CalendarTile.css';

const generateCircles = (colors: Color[], edge: number) => {
  const size = (edge * (1.3 + (colors.length - 2) * 0.1)) / colors.length;
  const shift = (edge - size) / (colors.length - 1);
  let position = -shift;
  return colors.map((color) => {
    position += shift;
    let colorVal = color;
    if (Object.keys(colorVal).length === 0) {
      colorVal = { color: '#dbdbdb', id: 'empty_color' };
    }
    return (
      <StyledCircle
        key={color.id}
        position={position}
        alone={colors.length === 1}
        color={colorVal.color}
        size={size}
      />
    );
  });
};

type TileProps = {
  colors?: Color[];
  day: number;
  handleDayChange: Function;
};

const CalendarTile = ({ colors, day, handleDayChange }: TileProps) => {
  const {
    gridTile: { size },
  } = useThemeContext();
  const [width] = useWindowSize();

  return (
    <StyledTile onClick={handleDayChange as MouseEventHandler<HTMLButtonElement>}>
      {day === -1 ? null : (
        <>
          {colors && generateCircles(colors, width > 768 ? size * 2 : size)}
          <StyledDay data-testid="day">{day}</StyledDay>
        </>
      )}
    </StyledTile>
  );
};

export default CalendarTile;
