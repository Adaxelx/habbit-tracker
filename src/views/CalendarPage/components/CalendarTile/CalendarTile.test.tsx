import React from 'react';
import 'jest-styled-components';
import { TestUtil, Color } from 'utils';
import CalendarTile from './CalendarTile';

describe('CalendarTile', () => {
  let util: TestUtil;

  let props: { colors: Color[]; day: number };

  beforeEach(() => {
    jest.clearAllMocks();

    props = {
      colors: [{ color: '#fff', id: 1 }],
      day: 1,
    };
    util = new TestUtil(<CalendarTile {...props} />);
  });

  it('should match snapshot', () => {
    expect(util.get('day').textContent).toEqual('1');
    expect(util.render.asFragment()).toMatchSnapshot();
  });

  it('it should render empty tile when day is -1', () => {
    props = {
      colors: [],
      day: -1,
    };
    util = new TestUtil(<CalendarTile {...props} />);
    expect(util.render.asFragment()).toMatchSnapshot();
  });

  it('it change styles when two colors appear', () => {
    props = {
      colors: [
        { color: '#fff', id: 1 },
        { color: '#ddd', id: 2 },
      ],
      day: 1,
    };
    util = new TestUtil(<CalendarTile {...props} />);
    expect(util.render.asFragment()).toMatchSnapshot();
  });
});
