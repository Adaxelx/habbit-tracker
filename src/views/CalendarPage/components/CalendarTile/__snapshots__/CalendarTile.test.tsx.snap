// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalendarTile it change styles when two colors appear 1`] = `
<DocumentFragment>
  .c0 {
  border: none;
  width: 50px;
  height: 50px;
  position: relative;
  overflow: hidden;
  background-color: #FFFFFF;
}

.c0:hover {
  cursor: pointer;
}

.c1 {
  position: absolute;
  border-radius: 50%;
  width: 65px;
  height: 65px;
  background-color: #fff;
  top: 0px;
  left: 0px;
}

.c2 {
  position: absolute;
  border-radius: 50%;
  width: 65px;
  height: 65px;
  background-color: #ddd;
  top: 35px;
  left: 35px;
}

.c3 {
  position: absolute;
  top: 50%;
  left: 50%;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
  z-index: 100;
}

@media (min-width:768px) {
  .c0 {
    width: 100px;
    height: 100px;
  }
}

<button
    class="c0"
  >
    <div
      class="c1"
      color="#fff"
      size="65"
    />
    <div
      class="c2"
      color="#ddd"
      size="65"
    />
    <p
      class="c3"
      data-testid="day"
    >
      1
    </p>
  </button>
</DocumentFragment>
`;

exports[`CalendarTile it should render empty tile when day is -1 1`] = `
<DocumentFragment>
  .c0 {
  border: none;
  width: 50px;
  height: 50px;
  position: relative;
  overflow: hidden;
  background-color: #FFFFFF;
}

.c0:hover {
  cursor: pointer;
}

@media (min-width:768px) {
  .c0 {
    width: 100px;
    height: 100px;
  }
}

<button
    class="c0"
  />
</DocumentFragment>
`;

exports[`CalendarTile should match snapshot 1`] = `
<DocumentFragment>
  .c0 {
  border: none;
  width: 50px;
  height: 50px;
  position: relative;
  overflow: hidden;
  background-color: #FFFFFF;
}

.c0:hover {
  cursor: pointer;
}

.c1 {
  position: absolute;
  border-radius: 50%;
  width: 92.3076923076923px;
  height: 92.3076923076923px;
  background-color: #fff;
  top: NaNpx;
  left: NaNpx;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
  top: 50%;
  left: 50%;
}

.c2 {
  position: absolute;
  top: 50%;
  left: 50%;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
  z-index: 100;
}

@media (min-width:768px) {
  .c0 {
    width: 100px;
    height: 100px;
  }
}

<button
    class="c0"
  >
    <div
      class="c1"
      color="#fff"
      size="120"
    />
    <p
      class="c2"
      data-testid="day"
    >
      1
    </p>
  </button>
</DocumentFragment>
`;
